llm/llama.go

```diff
@@ -59,104 +51,128 @@ func cudaVersion() (int, error) {
 
-	// Set OS specific llama.cpp runner paths
 	switch runtime.GOOS {
 	case "darwin":
-		// TODO: change to check metal version
-		llamaPath = osPath(path.Join("llama.cpp", runnerType, "build", "gpu", "bin"))
-		files = append(files, "ggml-metal.metal")
-	case "linux":
-		cudaVersion, err := cudaVersion()
-		if err != nil {
-			// fallback to CPU runner in the following the CUDA version check
-			log.Printf("failed to get CUDA version: %v", err)
+		files = []string{
+			path.Join(buildPath, "metal", "bin", "server"),
+			path.Join(buildPath, "metal", "bin", "ggml-metal.metal"),
+			path.Join(buildPath, "cpu", "bin", "server"),
 		}
-
-		switch cudaVersion {
-		case 11, 12:
-			cudaDir := fmt.Sprintf("cuda-%d", cudaVersion)
-			llamaPath = osPath(path.Join("llama.cpp", runnerType, "build", cudaDir, "bin"))
-		default:
-			if cudaVersion != -1 {
-				// a valid version was returned but it is not supported
-				log.Printf("CUDA version %d not supported, falling back to CPU", cudaVersion)
-			}
-			llamaPath = cpuPath
+	case "linux":
+		files = []string{
+			path.Join(buildPath, "cuda-12", "bin", "server"),
+			path.Join(buildPath, "cuda-11", "bin", "server"),
+			path.Join(buildPath, "cpu", "bin", "server"),
 		}
 	case "windows":
 		// TODO: select windows GPU runner here when available
-		files = []string{"server.exe"}
+		files = []string{
+			path.Join(buildPath, "cpu", "bin", "Release", "server.exe"),
+		}
 	default:
 		log.Printf("unknown OS, running on CPU: %s", runtime.GOOS)
-	}
-
-	// check if the runner exists, if not fallback to CPU runner
-	if _, err := fs.Stat(llamaCppEmbed, llamaPath); err != nil {
-		// fallback to CPU runner
-		llamaPath = cpuPath
-		files = []string{"server"}
-		if _, err := fs.Stat(llamaCppEmbed, llamaPath); err != nil {
-			log.Fatalf("llama.cpp executable not found")
+		files = []string{
+			path.Join(buildPath, "cpu", "bin", "server"),
 		}
-		log.Printf("llama.cpp %s executable not found, falling back to cpu", runnerType)
 	}
 
 	// copy the files locally to run the llama.cpp server
+	tmpDir, err := os.MkdirTemp("", "llama-*")
+	if err != nil {
+		log.Fatalf("load llama runner: failed to create temp dir: %v", err)
+	}
+	runnerAvailable := false // if no runner files are found in the embed, this flag will cause a fast fail
 	for _, f := range files {
-		srcPath := path.Join(llamaPath, f)
-		destPath := filepath.Join(tmpDir, f)
+		if _, err := fs.Stat(llamaCppEmbed, f); err != nil {
+			// this is expected, ollama may be compiled without all runners packed in
+			log.Printf("%s runner not found: %v", f, err)
+			continue
+		}
+		runnerAvailable = true
 
-		srcFile, err := llamaCppEmbed.Open(srcPath)
+		srcFile, err := llamaCppEmbed.Open(f)
 		if err != nil {
-			log.Fatalf("read llama.cpp %s: %v", f, err)
+			log.Fatalf("read llama runner %s: %v", f, err)
 		}
 		defer srcFile.Close()
 
+		// create the directory in case it does not exist
+		destPath := filepath.Join(tmpDir, strings.TrimPrefix(f, buildPath))
```

Comment:
Is it necessary to preserve part of the path? I would imagine `filepath.Dir` would be sufficient for most cases.